fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/ProgressiveLoanInterestRefundServiceImpl.java

```diff
@@ -76,52 +72,77 @@ private static void simulateRepaymentForDisbursements(LoanTransaction lt, final
         Loan loan = loanAssembler.assembleFrom(loanId);
-        BigDecimal totalInterestBeforeRefund = totalInterest(loan, BigDecimal.ZERO, relatedRefundTransactionDate);
-        BigDecimal totalInterestAfterRefund = totalInterest(loan, relatedRefundTransactionAmount, relatedRefundTransactionDate);
-        return totalInterestBeforeRefund.subtract(totalInterestAfterRefund);
+        if (processor == null) {
+            processor = loan.getTransactionProcessor();
+        }
+        if (!(processor instanceof AdvancedPaymentScheduleTransactionProcessor)) {
+            throw new IllegalArgumentException(
+                    "Wrong processor implementation. ProgressiveLoanInterestRefundServiceImpl requires AdvancedPaymentScheduleTransactionProcessor");
+        }
+
+        final AtomicReference<BigDecimal> refundFinal = new AtomicReference<>(BigDecimal.ZERO);
+
+        List<LoanTransaction> transactionsToReprocess = new ArrayList<>();
+        List<LoanTransactionType> interestRefundTypes = loan.getSupportedInterestRefundTransactionTypes();
+
+        var oldTransactions = loan.getLoanTransactions().stream() //
+                .filter(LoanTransaction::isNotReversed) //
+                .filter(lt -> !lt.isAccrual() && !lt.isAccrualActivity() && !lt.isInterestRefund()) //
+                .filter(lt -> oldTransactionIds.contains(lt.getId())).toList();
+
+        oldTransactions.stream().filter(lt -> interestRefundTypes.contains(lt.getTypeOf())) //
+                .forEach(t -> refundFinal.set(refundFinal.get().add(t.getAmount()))); //
+
+        newTransactions.stream() //
+                .filter(LoanTransaction::isNotReversed) //
+                .filter(lt -> interestRefundTypes.contains(lt.getTypeOf())) //
+                .forEach(t -> refundFinal.set(refundFinal.get().add(t.getAmount()))); //
+
+        oldTransactions.stream().filter(lt -> oldTransactionIds.contains(lt.getId()))
+                .filter(loanTransaction -> !interestRefundTypes.contains(loanTransaction.getTypeOf())) //
+                .forEach(lt -> simulateRepaymentForDisbursements(lt, refundFinal, transactionsToReprocess)); //
+
+        newTransactions.stream() //
+                .filter(LoanTransaction::isNotReversed) //
+                .filter(lt -> !lt.isAccrual() && !lt.isAccrualActivity() && !lt.isInterestRefund()) //
+                .filter(loanTransaction -> !interestRefundTypes.contains(loanTransaction.getTypeOf())) //
+                .map(LoanTransaction::copyTransactionProperties)
+                .forEach(lt -> simulateRepaymentForDisbursements(lt, refundFinal, transactionsToReprocess)); //
+
+        return recalculateTotalInterest((AdvancedPaymentScheduleTransactionProcessor) processor, loan, relatedRefundTransactionDate,
+                transactionsToReprocess);
+    }
+
+    @Override
+    public Money getTotalInterestRefunded(List<LoanTransaction> loanTransactions, Money zero) {
```

Comment:
Would you mind to remove the 2nd parameter? We should really not pass a "zero" money object... Would you mind rather pass the currency? It would be a little bit nicer...

Reply:
agree, done.